src/sentry/issues/escalating.py

```diff
@@ -71,66 +72,127 @@ def query_groups_past_counts(groups: Sequence[Group]) -> List[GroupsCountRespons
-    group_ids_by_project = _extract_project_and_group_ids(groups)
-    proj_ids, group_ids = [], []
-    processed_projects = 0
-    total_projects_count = len(group_ids_by_project)
-    organization_id = groups[0].project.organization.id
-
-    # This iteration guarantees that all groups for a project will be queried in the same call
-    # and only one page where the groups could be mixed with groups from another project
-    # Iterating over the sorted keys guarantees results for tests
-    for proj_id in sorted(group_ids_by_project.keys()):
-        _group_ids = group_ids_by_project[proj_id]
-        # Add them to the list of projects and groups to query
-        proj_ids.append(proj_id)
-        group_ids += _group_ids
-        processed_projects += 1
-        potential_num_elements = len(_group_ids) * BUCKETS_PER_GROUP
-        # This is trying to maximize the number of groups on the first page
-        if (
-            processed_projects < total_projects_count
-            and potential_num_elements < ELEMENTS_PER_SNUBA_PAGE
-        ):
-            continue
-
-        # TODO: Write this as a dispatcher type task and fire off a separate task per proj_ids
-        all_results += _query_with_pagination(
-            organization_id, proj_ids, group_ids, start_date, end_date
-        )
-        # We're ready for a new set of projects and ids
-        proj_ids, group_ids = [], []
+    processor = GroupCountProcessor(category=GroupCategory.ERROR)
+    all_results += processor.query_groups_past_counts(groups)
 
     return all_results
 
 
-def _query_with_pagination(
-    organization_id: int,
-    project_ids: Sequence[int],
-    group_ids: Sequence[int],
-    start_date: datetime,
-    end_date: datetime,
-) -> List[GroupsCountResponse]:
-    """Query Snuba for event counts for the given list of project ids and groups ids in
-    a time range."""
-    all_results = []
-    offset = 0
-    while True:
-        query = _generate_query(project_ids, group_ids, offset, start_date, end_date)
-        request = Request(
-            dataset=Dataset.Events.value,
-            app_id=REFERRER,
-            query=query,
-            tenant_ids={"referrer": REFERRER, "organization_id": organization_id},
+class GroupCountProcessor:
+    """Instanciate this class with a group category and then call query_groups_past_counts to
+    query Snuba for the counts for every group bucketed into hours."""
+
+    def __init__(self, category: GroupCategory) -> None:
+        self.category = category
+        self.start_date, self.end_date = _start_and_end_dates()
+
+    def query_groups_past_counts(self, groups: Sequence[Group]) -> List[GroupsCountResponse]:
+        """Given a list of groups, query Snuba for their hourly bucket count."""
+        all_results = []  # type: ignore[var-annotated]
+        if not groups:
+            return all_results
+
+        group_ids_by_project = self._extract_project_and_group_ids(groups)
```

Comment:
If you use the side by side diff and hide white spaces you will see that is mostly the same code up to here:
<img width="1278" alt="image" src="https://user-images.githubusercontent.com/44410/235511533-4dd47faa-bffb-4e9b-b8d1-e8b0b883115b.png">